velox/expression/SimpleFunctionAdapter.h

```diff
@@ -294,7 +312,33 @@ class SimpleFunctionAdapter : public VectorFunction {
           bits::setNull(nullBuffer, row);
         }
       };
-      if (allNotNull) {
+      if (callNullFree) {
+        // This results in some code duplication, but applying this check once
+        // per batch instead of once per row shows a significant performance
+        // improvement when there are no nulls.
+        if (applyContext.mayHaveNullsRecursive) {
+          applyContext.applyToSelectedNoThrow([&](auto row) INLINE_LAMBDA {
+            auto out = typename return_type_traits::NativeType();
+            auto containsNull = (readers.containsNull(row) || ...);
```

Comment:
One thing to consider if you decided not to merge the two functions
setChildrenMayHaveNulls and  mayHaveNullsRecursive is to call setChildrenMayHaveNulls 
only when applyContext.mayHaveNullsRecursive is true. 
(readers.setChildrenMayHaveNulls(),...);

Reply:
If I do that won't I have to compute mayHaveNullsRecursive twice, once to compute the value of applyContext.mayHaveNullsRecursive, and once again to run (readers.setChildrenMayHaveNulls(),...); if it's true?

Reply:
decided not to merge  refers to my previous comment about caching while running mayHaveNullsRecursive(). I was saying id you decided not to do that to make the caching explicit. would it make sense to move the call to setChildrenMayHaveNulls here?

Reply:
so basically having it here means we do not need to run setChildrenMayHaveNulls() when for the path when is applyContext.mayHaveNullsRecursive is false.

Reply:
Where it is now, the code looks like this

      if constexpr (FUNC::udf_has_callNullFree) {
        oneReader.setChildrenMayHaveNulls();
        applyContext.mayHaveNullsRecursive |= oneReader.mayHaveNullsRecursive();
      }

We use the cached value computed in oneReader.setChildrenMayHaveNulls() in the next line when we call oneReader.mayHaveNullsRecursive()

If I don't call setChildrenMayHaveNulls there, I'll have to do the same computation in mayHaveNullsRecursive (except in the case where the top level value may be null).  And then I'll have to do it again when I call setChildrenMayHaveNulls here.